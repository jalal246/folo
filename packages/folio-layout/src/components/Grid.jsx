import React from "react";
import PropTypes from "prop-types";

import {
  GRID,
  STRETCH,
  CENTER,
  SPACE_BETWEEN,
  AUTO,
  FR,
  AUTO_FIT,
  DEFAULT_GAP
} from "../constants";

const propTypes = {
  /**
   * custom render-component
   */
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

  /**
   * number of columns in grid
   */
  col: PropTypes.number,
  colWidth: PropTypes.string,
  colMinWidth: PropTypes.string,
  colMaxWidth: PropTypes.string,

  /**
   * number of rows in grid
   */
  row: PropTypes.number,
  rowWidth: PropTypes.string,
  rowMinWidth: PropTypes.string,
  rowMaxWidth: PropTypes.string,

  style: PropTypes.objectOf(PropTypes.string),
  //
  children: PropTypes.node.isRequired
};

const defaultProps = {
  component: "div",

  col: null,
  colWidth: undefined,
  colMinWidth: AUTO,
  colMaxWidth: FR,

  row: null,
  rowWidth: undefined,
  rowMinWidth: AUTO,
  rowMaxWidth: FR,
  style: {}
};

/**
 * call repeat() CSS function
 * depending on length, min,max
 * with some enhancements
 *
 *
 * @param {number} length row or column number in grid
 * @param {string} min minimum unit for row or column
 * @param {string} max maximum unit for row or column
 * @return {string}   repeat() CSS function represents a repeated fragment of the track list
 */
function repeat(length, fixed, min, max) {
  return `repeat(${length}, ${fixed || `minmax(${min}, ${max})`})`;
}

class Grid extends React.PureComponent {
  render() {
    const {
      component: CellComponent,

      col,
      colWidth,
      colMinWidth,
      colMaxWidth,

      row,
      rowWidth,
      rowMinWidth,
      rowMaxWidth,

      isCenter,

      // TODO: add style and shape to propTypes
      style: {
        display = GRID,

        alignItems = isCenter ? CENTER : STRETCH,

        // eslint-disable-next-line
        justifyContent = col || colWidth
          ? SPACE_BETWEEN
          : isCenter
            ? CENTER
            : STRETCH,

        gap = DEFAULT_GAP,

        ...otherStyles
      },
      children,

      //
      ...otherProps
    } = this.props;

    const style = {
      display,

      ...(row
        ? { gridTemplateRows: repeat(row, rowWidth, rowMinWidth, rowMaxWidth) }
        : rowWidth && { gridAutoRows: rowWidth }),

      ...(colWidth && !col
        ? { gridAutoColumns: colWidth }
        : {
            gridTemplateColumns: repeat(
              col || AUTO_FIT,
              colWidth,
              colMinWidth,
              colMaxWidth
            )
          }),

      alignItems,
      justifyContent,
      gap,
      ...otherStyles
    };

    return (
      <CellComponent style={style} {...otherProps}>
        {children}
      </CellComponent>
    );
  }
}

Grid.propTypes = propTypes;
Grid.defaultProps = defaultProps;

export default Grid;
