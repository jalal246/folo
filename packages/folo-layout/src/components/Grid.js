import React from "react";

import {
  GRID,
  STRETCH,
  CENTER,
  SPACE_BETWEEN,
  AUTO,
  FR,
  AUTO_FIT,
  DEFAULT_GAP,
} from "../constants";

/**
 * call repeat() CSS function
 * depending on length, min,max
 * with some enhancements
 *
 *
 * @param {number} length row or column number in grid
 * @param {string} min minimum unit for row or column
 * @param {string} max maximum unit for row or column
 * @return {string}   repeat() CSS function represents a repeated fragment of the track list
 */
function repeat(length, fixed, min, max) {
  return `repeat(${length}, ${fixed || `minmax(${min}, ${max})`})`;
}

const Grid = (props) => {
  const {
    component: CellComponent = "div",

    col,
    colWidth,
    colMinWidth = AUTO,
    colMaxWidth = FR,

    row,
    rowWidth,
    rowMinWidth,
    rowMaxWidth,

    isCenter = false,

    style: {
      display = GRID,

      alignItems = isCenter ? CENTER : STRETCH,

      // eslint-disable-next-line
      justifyContent = col || colWidth
        ? SPACE_BETWEEN
        : isCenter
        ? CENTER
        : STRETCH,

      gap = DEFAULT_GAP,

      ...otherStyles
    } = {},
    children,

    //
    ...rest
  } = props;

  const style = {
    display,

    ...(row
      ? { gridTemplateRows: repeat(row, rowWidth, rowMinWidth, rowMaxWidth) }
      : rowWidth && { gridAutoRows: rowWidth }),

    ...(colWidth && !col
      ? { gridAutoColumns: colWidth }
      : {
          gridTemplateColumns: repeat(
            col || AUTO_FIT,
            colWidth,
            colMinWidth,
            colMaxWidth
          ),
        }),

    alignItems,
    justifyContent,
    gap,
    ...otherStyles,
  };

  return (
    <CellComponent style={style} {...rest}>
      {children}
    </CellComponent>
  );
};

export default Grid;
