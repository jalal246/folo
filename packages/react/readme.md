# React Components

> Folo React components

Directory content:

// TODO
