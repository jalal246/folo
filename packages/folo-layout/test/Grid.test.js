import React from "react";
import { render, screen, cleanup } from "@testing-library/react";

import Grid from "../src/components/Grid";

const TEST_ID = "testForGrid";

const DEFAULT_DISPLAY = "grid";
const STRETCH = "stretch";
const DEFAULT_GAP = "1em";
const SPACE_BETWEEN = "space-between";
const AUTO_FIT = "auto-fit";
const CENTER = "center";

const MyGrid = (props = {}) => (
  <Grid data-testid={TEST_ID} {...props}>
    <div />
  </Grid>
);

describe("Grid - Explicit Layout", () => {
  afterEach(() => {
    cleanup();
  });

  describe("Testing Style", () => {
    it("Returns default style", () => {
      render(<MyGrid />);

      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: `repeat(${AUTO_FIT}, minmax(auto, 1fr))`,
        alignItems: STRETCH,
        justifyContent: STRETCH,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns default style", () => {
      const DISPLAY_INLINE = "inline-grid";
      render(<MyGrid style={{ display: DISPLAY_INLINE }} />);

      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_INLINE,
        gridTemplateColumns: `repeat(${AUTO_FIT}, minmax(auto, 1fr))`,
        alignItems: STRETCH,
        justifyContent: STRETCH,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("Testing alignItems", () => {
    it("Returns alignItems justifyContent center when no columns passed", () => {
      render(<MyGrid isCenter />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: `repeat(${AUTO_FIT}, minmax(auto, 1fr))`,
        alignItems: CENTER,
        justifyContent: CENTER,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns alignItems center and justifyContent when columns passed", () => {
      render(<MyGrid isCenter col={1} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: "repeat(1, minmax(auto, 1fr))",
        alignItems: CENTER,
        justifyContent: SPACE_BETWEEN,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("Testing rows", () => {
    it("Returns gridTemplateRows when passing row with minmax", () => {
      render(<MyGrid row={1} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: `repeat(${AUTO_FIT}, minmax(auto, 1fr))`,
        gridTemplateRows: "repeat(1, minmax(auto, 1fr))",
        alignItems: STRETCH,
        justifyContent: STRETCH,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns gridTemplateRows when passing row with width", () => {
      render(<MyGrid row={1} rowWidth="20px" />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: `repeat(${AUTO_FIT}, minmax(auto, 1fr))`,
        gridTemplateRows: "repeat(1, 20px)",
        alignItems: STRETCH,
        justifyContent: STRETCH,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns gridAutoRows when passing rowWidth only", () => {
      render(<MyGrid rowWidth="20px" />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: `repeat(${AUTO_FIT}, minmax(auto, 1fr))`,
        gridAutoRows: "20px",
        alignItems: STRETCH,
        justifyContent: STRETCH,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("Testing columns", () => {
    it("Returns gridAutoColumns when passing colWidth only", () => {
      render(<MyGrid colWidth="20px" />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridAutoColumns: "20px",
        alignItems: STRETCH,
        justifyContent: SPACE_BETWEEN,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns gridTemplateColumns when passing col", () => {
      render(<MyGrid col={4} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: "repeat(4, minmax(auto, 1fr))",
        alignItems: STRETCH,
        justifyContent: SPACE_BETWEEN,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns gridTemplateColumns when passing colWidth and col", () => {
      render(<MyGrid col={4} colWidth="20px" />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DEFAULT_DISPLAY,
        gridTemplateColumns: "repeat(4, 20px)",
        alignItems: STRETCH,
        justifyContent: SPACE_BETWEEN,
        gap: DEFAULT_GAP,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });
});
