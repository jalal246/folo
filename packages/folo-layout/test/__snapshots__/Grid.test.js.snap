// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid - Explicit Layout Testing Style Returns default style 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing Style Returns default style 2`] = `undefined`;

exports[`Grid - Explicit Layout Testing alignItems Returns alignItems center and justifyContent when columns passed 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing alignItems Returns alignItems justifyContent center when no columns passed 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing columns Returns gridAutoColumns when passing colWidth only 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing columns Returns gridTemplateColumns when passing col 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing columns Returns gridTemplateColumns when passing colWidth and col 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing rows Returns gridAutoRows when passing rowWidth only 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing rows Returns gridTemplateRows when passing row with minmax 1`] = `undefined`;

exports[`Grid - Explicit Layout Testing rows Returns gridTemplateRows when passing row with width 1`] = `undefined`;
